### PR TITLE
Cardinality and Quotients of Finite Sets

### PR DESCRIPTION
**Pull request by kangrongji**
This PR is a following-up of #620 and it contains:

- Many properties of cardinality, e.g., facts about cardinality, some basic counting formula and a general version of pigeonhole principle;
- A universe polymorphic version of `Fin` and based on that an inductive eliminator of `FinSet` towards propositions, making certain proof cleaner;
- Facts about quotients. In particular a proof that quotient of finite set by decidable equivalence relation is still finite set;
- Modifications in library.

See `Cubical.Experiments.Combinatorics` that things actually work and compute.

### CODE REVIEW

**Comment by kangrongji** [comment]
I guess I should open another branch for this PR...

File: Cubical/Data/FinSet/Cardinality.agda [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.HLevels
+open import Cubical.Foundations.Function
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.Univalence
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Equiv.Properties
+
+open import Cubical.HITs.PropositionalTruncation hiding (elim')
+open import Cubical.HITs.SetTruncation renaming (rec to SetRec ; elim to SetElim) hiding (rec2)
```
Comment (ecavallo) [inline Cubical/Data/FinSet/Cardinality.agda:19]: I suggest using `open import Cubical.HITs.SetTruncation as Set` and then you can write `Set.rec` and `Set.elim`, same for `PropositionalTruncation` and in other files.

File: Cubical/Data/FinSet/Induction.agda [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    elimProp' : (X : FinSet ℓ) → P X
+    elimProp' = elimProp elimProp𝔽in
+
+  module _
+    (p0 : P 𝟘)(p1 : P 𝟙)
+    (p+ : {X Y : FinSet ℓ} → P X → P Y → P (X + Y)) where
+
+    elimProp'' : (X : FinSet ℓ) → P X
+    elimProp'' = elimProp' p0 (λ p → p+ p1 p)
```
Comment (ecavallo) [inline Cubical/Data/FinSet/Induction.agda:131]: How about calling these `elimProp𝟙+` and `elimProp+`? Just to be a bit more descriptive.

File: Cubical/Data/FinSet/Quotients.agda [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+open Iso
+
+module _
+  (ℓ : Level) where
+
+  DecProp : Type (ℓ-suc ℓ)
+  DecProp = Σ[ P ∈ hProp ℓ ] Dec (P .fst)
+
+  isSetDecProp : isSet DecProp
+  isSetDecProp = isOfHLevelΣ 2 isSetHProp (λ P → isProp→isSet (isPropDec (P .snd)))
```
Comment (ecavallo) [inline Cubical/Data/FinSet/Quotients.agda:37]: Could these be moved to where `Dec` is defined?

File: Cubical/Data/Nat/Order.agda [hunk 4]
```diff
@@ -92,6 +98,9 @@ pred-≤-pred (k , p) = k , injSuc ((sym (+-suc k _)) ∙ p)
   l3 : 0 ≡ i
   l3 = sym (snd (m+n≡0→m≡0×n≡0 l2))
 
+≤-+ : m ≤ n → l ≤ k → m + l ≤ n + k
+≤-+ p q = ≤-trans (≤-+k p) (≤-k+ q)
+
```
Comment (ecavallo) [inline Cubical/Data/Nat/Order.agda:103]: How about renaming this to `≤-+-≤` to match `<-+-≤`?

File: Cubical/Data/Nat/Order.agda [hunk 5]
```diff
@@ -157,6 +166,9 @@ predℕ-≤-predℕ {suc m} {suc n} ineq = pred-≤-pred ineq
 +-<-+ : m < n → k < l → m + k < n + l
```
Comment (ecavallo) [inline Cubical/Data/Nat/Order.agda:166]: Rename to `<-+-<`?

File: Cubical/Data/SumFin/Properties.agda [hunk 6]
```diff
@@ -90,3 +91,20 @@ SumFin∥∥≃ 0 = propTruncIdempotent≃ (isProp⊥)
...
+
+Iso-⊤⊎⊤-Bool : Iso (⊤ ⊎ ⊤) Bool
+Iso-⊤⊎⊤-Bool .fun (inl tt) = true
+Iso-⊤⊎⊤-Bool .fun (inr tt) = false
+Iso-⊤⊎⊤-Bool .inv true = inl tt
+Iso-⊤⊎⊤-Bool .inv false = inr tt
+Iso-⊤⊎⊤-Bool .leftInv (inl tt) = refl
+Iso-⊤⊎⊤-Bool .leftInv (inr tt) = refl
+Iso-⊤⊎⊤-Bool .rightInv true = refl
+Iso-⊤⊎⊤-Bool .rightInv false = refl
```
Comment (ecavallo) [inline Cubical/Data/SumFin/Properties.agda:107]: Can this go somewhere in Data.Bool?

**Review by ecavallo — CHANGES_REQUESTED** [review]
This is great! I just have some clean-up requests before I merge.

File: Cubical/Data/Fin/Properties.agda [hunk 7]
```diff
@@ -634,3 +636,43 @@ FinData≃Fin N = isoToEquiv (FinDataIsoFin N)
...
+
+Fin≤1 : (n : ℕ) → n ≤ 1 → isProp (Fin n)
+Fin≤1 0 _ = isPropFin0
+Fin≤1 1 _ = isContr→isProp isContrFin1
+Fin≤1 (suc (suc n)) p = Empty.rec (¬-<-zero (pred-≤-pred p))
+
+propFin : (n : ℕ) → isProp (Fin n) → n ≤ 1
+propFin 0 _ = ≤-solver 0 1
+propFin 1 _ = ≤-solver 1 1
+propFin (suc (suc n)) p = Empty.rec (fzero≠fone (p fzero fone))
```
[the remark this answers is not in the document]
Comment (kangrongji) [reply Cubical/Data/Fin/Properties.agda:678]: done

File: Cubical/Data/Fin/Properties.agda [hunk 8]
```diff
@@ -634,3 +636,43 @@ FinData≃Fin N = isoToEquiv (FinDataIsoFin N)
 
 FinData≡Fin : (N : ℕ) → FinData N ≡ Fin N
 FinData≡Fin N = ua (FinData≃Fin N)
+
+-- propositional truncation of Fin
+
+∥Fin∥ : (n : ℕ) → Dec ∥ Fin n ∥
+∥Fin∥ 0 = no (∥∥rec isProp⊥ ¬Fin0)
+∥Fin∥ (suc n) = yes ∣ fzero ∣
```
[the remark this answers is not in the document]
Comment (kangrongji) [reply Cubical/Data/Fin/Properties.agda:644]: done

File: Cubical/Data/FinSet/Cardinality.agda (hunk 1 above)
Comment (kangrongji) [reply Cubical/Data/FinSet/Cardinality.agda:19]: done

File: Cubical/Data/FinSet/Cardinality.agda [hunk 9]
```diff
@@ -0,0 +1,620 @@
...
+
+  sum𝔽in1+n : sum (𝔽in (1 + n)) f ≡ f (inl tt*) + sum (𝔽in n) (f ∘ inr)
+  sum𝔽in1+n = sum⊎ 𝟙 (𝔽in n) f ∙ (λ i → sum𝟙 (f ∘ inl) i + sum (𝔽in n) (f ∘ inr))
+
+  prod𝔽in1+n : prod (𝔽in (1 + n)) f ≡ f (inl tt*) · prod (𝔽in n) (f ∘ inr)
+  prod𝔽in1+n = prod⊎ 𝟙 (𝔽in n) f ∙ (λ i → prod𝟙 (f ∘ inl) i · prod (𝔽in n) (f ∘ inr))
+
+_^_ : ℕ → ℕ → ℕ
+m ^ 0 = 1
+m ^ (suc n) = m · m ^ n
```
[the remark this answers is not in the document]
Comment (kangrongji) [reply Cubical/Data/FinSet/Cardinality.agda:256]: done

File: Cubical/Data/Nat/Order.agda (hunk 4 above)
Comment (kangrongji) [reply Cubical/Data/Nat/Order.agda:103]: done

File: Cubical/Data/Nat/Order.agda (hunk 5 above)
Comment (kangrongji) [reply Cubical/Data/Nat/Order.agda:166]: done

File: Cubical/Data/SumFin/Properties.agda (hunk 6 above)
Comment (kangrongji) [reply Cubical/Data/SumFin/Properties.agda:107]: done

File: Cubical/Data/FinSet/Induction.agda [hunk 10]
```diff
@@ -0,0 +1,131 @@
+{-
+
+Indutiive eliminators to directly prove properties of all finite sets
```
[the remark this answers is not in the document]
Comment (kangrongji) [reply Cubical/Data/FinSet/Induction.agda:3]: done

File: Cubical/Data/FinSet/Induction.agda (hunk 2 above)
Comment (kangrongji) [reply Cubical/Data/FinSet/Induction.agda:131]: done

File: Cubical/Data/FinSet/Properties.agda [hunk 11]
```diff
@@ -31,11 +32,25 @@ private
...
+  _⋆_ = compEquiv
+
+  _⋆̂_ : ∥ A ≃ B ∥ → ∥ B ≃ C ∥ → ∥ A ≃ C ∥
+  _⋆̂_ = rec2 isPropPropTrunc (λ p q → ∣ p ⋆ q ∣)
+
+module _
+  {A : Type ℓ}{B : Type ℓ'} where
+
+  ∣invEquiv∣ : ∥ A ≃ B ∥ → ∥ B ≃ A ∥
+  ∣invEquiv∣ = rec isPropPropTrunc (λ p → ∣ invEquiv p ∣)
```
[the remark this answers is not in the document]
Comment (kangrongji) [reply Cubical/Data/FinSet/Properties.agda:53]: done

File: Cubical/Data/FinSet/Quotients.agda (hunk 3 above)
Comment (kangrongji) [reply Cubical/Data/FinSet/Quotients.agda:37]: Done. I have to create a new file to resolve cyclic dependency, though.

Comment (kangrongji) [reply Cubical/Data/FinSet/Quotients.agda:37]: Done. I have to create a new file to resolve cyclic dependency, though.

File: Cubical/Data/FinSet/Properties.agda [hunk 12]
```diff
@@ -31,11 +32,25 @@ private
...
-_⋆_ = compEquiv
+module _
+  {A : Type ℓ}{B : Type ℓ'}{C : Type ℓ''} where
 
-infixr 30 _⋆_
+  infixr 30 _⋆_
+  infixr 30 _⋆̂_
+
+  _⋆_ : A ≃ B → B ≃ C → A ≃ C
+  _⋆_ = compEquiv
```
[the remark this answers is not in the document]
Comment (kangrongji) [reply Cubical/Data/FinSet/Properties.agda:44]: done

File: Cubical/Data/FinSet/Cardinality.agda [hunk 13]
```diff
@@ -0,0 +1,620 @@
...
+    elimProp
+      (λ X → (f : X .fst → ℕ)(c : ℕ)(h : (x : X .fst) → f x ≡ c) → prod X f ≡ c ^ (card X))
+      (λ X → isPropΠ3 (λ _ _ _ → isSetℕ _ _))
+      (λ n f c h → prodConst𝔽in n f c h ∙ (λ i → c ^ card𝔽in {ℓ = ℓ} n (~ i))) X f c h
+
+≡≤ : {m n l k r s : ℕ} → m ≤ n → l ≤ k → r ≡ m + l → s ≡ n + k → r ≤ s
+≡≤ {m = m} {l = l} {k = k} p q u v = subst2 (_≤_) (sym u) (sym v) (≤-+ p q)
+
+≡< : {m n l k r s : ℕ} → m < n → l ≤ k → r ≡ m + l → s ≡ n + k → r < s
+≡< {m = m} {l = l} {k = k} p q u v = subst2 (_<_) (sym u) (sym v) (<-+-≤ p q)
```
[the remark this answers is not in the document]
Comment (kangrongji) [reply Cubical/Data/FinSet/Cardinality.agda:294]: done

**Comment by kangrongji** [comment]
@ecavallo Thanks for reviewing! Everything is done now.

**Comment by ecavallo** [comment]
Thanks!

**Pull request merged**